### PR TITLE
Add command-line tool support for Lucene-based metadata storage

### PR DESCRIPTION
**Pull request by ywelsch**
Adds command-line tool support (unsafe-bootstrap, detach-cluster, repurpose, & shard commands) for the Lucene-based metadata storage.

Relates #48701

### CODE REVIEW

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/2

**Review by henningandersen — APPROVED** [review]
LGTM.

I left a few minor comments to consider. I assume the two important TODOs (setLastAcceptedTerm and dangling indices) will be tackled in follow-ups.

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 1]
```diff
@@ -72,19 +74,53 @@ public MetaData getMetaData() {
...
         assert persistedState.get() == null : "should only start once, but already have " + persistedState.get();
 
         if (DiscoveryNode.isMasterNode(settings) || DiscoveryNode.isDataNode(settings)) {
             try {
-                persistedState.set(lucenePersistedStateFactory.loadPersistedState((version, metadata) ->
-                    prepareInitialClusterState(transportService, clusterService,
+                LucenePersistedStateFactory.OnDiskState onDiskState = lucenePersistedStateFactory.loadBestOnDiskState();
+
+                if (onDiskState == LucenePersistedStateFactory.NO_ON_DISK_STATE) {
+                    assert Version.CURRENT.major <= Version.V_7_0_0.major + 1 :
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: We also return `LucenePersistedStateFactory.NO_ON_DISK_STATE` if starting up a brand-new node?

(I know this is merely moving code from an earlier PR, I missed this last time)

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I do not think we should expose the internals of `LucenePersistedStateFactory` as much as we are proposing here. The internal details and the state-loading logic are once again spilling into `GatewayMetaState` and elsewhere. I would prefer to keep things more isolated and add public methods to `LucenePersistedStateFactory` as needed to support the non-standard ways that the node tool accesses the on-disk state.

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: Sorry, ignore this comment, there's not really an obvious better way to assert this.

File: server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java [hunk 2]
```diff
@@ -130,28 +135,39 @@ private void processMasterNoDataNode(Terminal terminal, Path[] dataPaths) throws
...
 
-        Set<Path> indexPaths = uniqueParentPaths(shardDataPaths);
-        Set<String> indexUUIDs = indexUUIDsFor(indexPaths);
-        outputVerboseInformation(terminal, nodePaths, shardDataPaths, indexUUIDs);
+        final LucenePersistedStateFactory psf = createLucenePersistedStateFactory(dataPaths);
+
+        final MetaData metaData = loadClusterState(terminal, env, psf).metaData();
+
+        final Set<Path> indexPaths = uniqueParentPaths(shardDataPaths);
+        final Set<String> indexUUIDs = Sets.union(indexUUIDsFor(indexPaths),
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/env/NodeRepurposeCommand.java]: fixed in fd3dd91

File: server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java [hunk 3]
```diff
@@ -124,58 +121,35 @@
...
-            public void clean() {
-
-            }
-        });
-    }
-
-    public LucenePersistedStateFactory(NodeEnvironment nodeEnvironment, NamedXContentRegistry namedXContentRegistry, BigArrays bigArrays,
-                                       LegacyLoader legacyLoader) {
-        this.nodeEnvironment = nodeEnvironment;
+    public LucenePersistedStateFactory(Path[] dataPaths, String nodeId, boolean preserveUnknownCustoms,
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java]: adapted in cffdd8a.
The way I had it, the parameter list had the same suffix. Now it does not have same prefix nor suffix...

File: server/src/main/java/org/elasticsearch/cluster/coordination/ElasticsearchNodeCommand.java [hunk 4]
```diff
@@ -26,41 +26,77 @@
...
 import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public abstract class ElasticsearchNodeCommand extends EnvironmentAwareCommand {
-    private static final Logger logger = LogManager.getLogger(ElasticsearchNodeCommand.class);
+    protected static final Logger logger = LogManager.getLogger(ElasticsearchNodeCommand.class);
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/ElasticsearchNodeCommand.java]: adapted in 0c8bb67

File: server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java [hunk 5]
```diff
@@ -72,19 +74,53 @@ public MetaData getMetaData() {
...
+                        "legacy metadata loader is not needed anymore from v9 onwards";
+                    final Tuple<Manifest, MetaData> legacyState = metaStateService.loadFullState();
+                    if (legacyState.v1().isEmpty() == false) {
+                        onDiskState = new LucenePersistedStateFactory.OnDiskState(lucenePersistedStateFactory.getNodeId(),
+                            null, legacyState.v1().getCurrentTerm(),
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java]: adapted in 0c8bb67

File: server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java [hunk 6]
```diff
@@ -224,14 +184,14 @@ Directory createDirectory(Path path) throws IOException {
...
 
-    private static class OnDiskState {
-        final String nodeId;
-        final Path dataPath;
-        final long currentTerm;
-        final long lastAcceptedVersion;
-        final MetaData metaData;
+    public static class OnDiskState {
+        public final String nodeId;
+        public final Path dataPath;
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/gateway/LucenePersistedStateFactory.java]: adapted in 0c8bb67

**Review by ywelsch — COMMENTED** [review]
> I do not think we should expose the internals of LucenePersistedStateFactory as much as we are proposing here.

I've hidden more things away in 0c8bb67. I don't want to go back to `LegacyLoader` and introducing more callbacks to `LucenePersistedStateFactory` (e.g. to adapt cluster state). I think that as it stands this class has a minimal interface, and does not need to concern itself with legacy stuff.

> I would prefer to keep things more isolated and add public methods to LucenePersistedStateFactory as needed to support the non-standard ways that the node tool accesses the on-disk state.

That was my starting point, but after adding more bells and whistles to `LucenePersistedStateFactory` that were essentially irrelevant for what this class was trying to achieve, I didn't like it and got to the current solution.

I find the current minimal Writer/OnDiskState interfaces a good compromise between having a closed interface and unnecessary parameterization of the class.

**Review by DaveCTurner — APPROVED** [review]
LGTM thanks for the re-organisation.

**Pull request merged**